File: 20221002_Sunday/20221002.go
// Leetcode: 1155. Number of Dice Rolls With Target Sum.
// https://leetcode.com/problems/number-of-dice-rolls-with-target-sum/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 24 ms, faster than 61.36% of Go online submissions for Number of Dice Rolls With Target Sum.
// Memory Usage: 5.3 MB, less than 72.73% of Go online submissions for Number of Dice Rolls With Target Sum.
func numRollsToTarget(n int, k int, target int) int {
  dp := make([][]int,n)
  mod := 1000000007
  for i,_ := range(dp){
    dp[i] = make([]int, target+1)
  }
  for col := 1; col <= target; col +=1{
    if col > k {break}
    dp[0][col] = 1
  }
  for row := 1; row < n; row +=1{
    for col := 1; col <= target; col +=1 {
      for face := 1; face <= k; face += 1{
        if (0 > col - face) {continue}
        dp[row][col] = (dp[row][col] + dp[row-1][col-face]) % mod
      }
    }
  }
  return dp[n-1][target]
}

File: 20221002_Sunday/20221002.rb
# Leetcode: 1155. Number of Dice Rolls With Target Sum.
# https://leetcode.com/problems/number-of-dice-rolls-with-target-sum/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 690 ms, faster than 87.50% of Ruby online submissions for Number of Dice Rolls With Target Sum.
# Memory Usage: 211.8 MB, less than 100.00% of Ruby online submissions for Number of Dice Rolls With Target Sum.
# @param {Integer} n
# @param {Integer} k
# @param {Integer} target
# @return {Integer}
def num_rolls_to_target(n, k, target)
  # Reach row is a dice.
  # Each column is a value.
  dp = Array.new(n).map{ Array.new(target+1,0)}
  mod = (10**9) + 7

  # Fill first row.
  # There is only one way to fill the column.
  # But only if there is a face on a dice for this collumn.
  (1..target).each do |col|
    break if col > k
    dp[0][col] = 1  
  end
  # Get data from the previous dice. From the previous row.
  # Current is added to previous.
  (1...n).each do |row|
    (1..target).each do |col|
      (1..k).each do |face|
        next if 0 > col - face # Fixed here.
        dp[row][col] = (dp[row][col] + dp[row-1][col - face]) % mod
      end
    end
  end
  dp[n-1][target]
end
